backend/routers/transcribe.py

```diff
@@ -87,39 +93,74 @@ async def _websocket_util(
         await websocket.close()
         return
 
-    vad_iterator = VADIterator(model, sampling_rate=sample_rate)  # threshold=0.9
+    threshold = 0.6 # Currently most fitting threshold
+    vad_iterator = VADIterator(model, sampling_rate=sample_rate, threshold=threshold) 
     window_size_samples = 256 if sample_rate == 8000 else 512
+    if codec == 'opus':
+        decoder = opuslib.Decoder(sample_rate, channels)
 
     async def receive_audio(socket1, socket2):
-        # audio_buffer = bytearray()
+        nonlocal is_speech_active, last_speech_time, decoder
+        audio_buffer = deque(maxlen=sample_rate * 1)  # 1 secs
+        databuffer = bytearray(b"")
+        
+        REALTIME_RESOLUTION = 0.01
+        if codec == 'pcm8':
+            sample_width = 1
+        else:
+            sample_width = 2
+        if sample_width:
+            byte_rate = sample_width * sample_rate * channels
+            chunk_size = int(byte_rate * REALTIME_RESOLUTION)
+        else:
+            chunk_size = 4096 # Arbitrary value
+            
         timer_start = time.time()
         try:
+            sample_width = 1 if codec == "pcm8" else 2
             while True:
                 data = await websocket.receive_bytes()
-                # print(len(data))
-                # audio_buffer.extend(data)
+                if codec == 'opus':
+                    data = decoder.decode(data, frame_size=320) # 160 if want lower latency
+                    # audio = AudioSegment(data=data, sample_width=sample_width, frame_rate=sample_rate, channels=channels, format='opus')
+                    # samples = torch.tensor(audio.get_array_of_samples()).float() / 32768.0
+                    samples = torch.frombuffer(data, dtype=torch.int16).float() / 32768.0
+                elif codec in ['pcm8', 'pcm16']:
+                    dtype = torch.int8 if codec == 'pcm8' else torch.int16
+                    writeable_data = bytearray(data)
+                    samples = torch.frombuffer(writeable_data, dtype=dtype).float()
+                    samples = samples / (128.0 if codec == 'pcm8' else 32768.0)
```

Comment:
@0xzre here is a slightly simpler approach (some of this can be combined), but once the audio has been extracted you can use the following code to process it whether it's opus with 16Khz sample rate or pcm8 with 8Khz sample rate (I haven't tested with pcm16):

```
    def convert_audio_bytes_to_resampled_numpy(self, audio_bytes: bytes):
        audio_data = np.frombuffer(audio_bytes, dtype=np.int16).astype(np.float32) / 32767.0

        # Normalize the audio to bring the peak to 1.0
        max_val = np.max(np.abs(audio_data))
        if max_val > 0:
            audio_data = audio_data / max_val

        waveform = torch.from_numpy(audio_data).unsqueeze(0)
        
        # Ensure the audio is mono
        if waveform.shape[0] > 1:
            waveform = waveform.mean(dim=0, keepdim=True)
        
        # Resample if the source sample rate 8Khz
        if self.source_sample_rate == 8000:
            waveform = torchaudio.functional.resample(waveform, orig_freq=self.source_sample_rate, new_freq=16000)
        
        resampled_np = waveform.squeeze().numpy()

        return resampled_np
```

Also as a side note I found that the audio is really quiet and so normalizing it seems to help bring it up a little bit without added artifacts to the audio. However, if you want it even louder (which may help with VAD) you can also add this:

```
gain_factor = 1.2
audio_data = np.clip(audio_data * gain_factor, -1.0, 1.0)
```

Reply:
Thanks! I get the idea of your code and agree but I think normalizing naively is not that good for VAD because. Still will try normalizing it though to see if it's better